File: bindings/rust/examples/concurrent_demo.rs
use sop::{Context, Database, DatabaseOptions, Item, L2CacheType, DatabaseType};
use std::sync::Arc;
use std::thread;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

fn main() {
    println!("--- Concurrent Transactions Demo (Standalone) ---");
    println!("Demonstrating multi-threaded access without client-side locks.");
    
    // 1. Initialize Context
    // We wrap it in Arc to share across threads.
    let ctx = Arc::new(Context::new());
    if let Some(err) = ctx.error() {
        eprintln!("Error creating context: {}", err);
        return;
    }

    // 2. Configure Database
    let mut options = DatabaseOptions::default();
    options.stores_folders = Some(vec!["sop_data_concurrent".to_string()]);
    options.db_type = DatabaseType::Standalone;
    options.cache_type = L2CacheType::InMemory;
    
    let db = match Database::new(&ctx, options) {
        Ok(db) => db,
        Err(e) => {
            eprintln!("Error creating database: {}", e);
            return;
        }
    };
    println!("Database created with ID: {}", db.id);

    // 3. Seed the B-Tree
    // IMPORTANT: Pre-seed the B-Tree with one item to establish the root node.
    {
        let trans = db.begin_transaction(&ctx).unwrap();
        let btree = db.new_btree::<i64, String>(&ctx, "concurrent_tree", &trans, None).unwrap();
        btree.add(&ctx, -1, "Root Seed Item".to_string()).unwrap();
        trans.commit(&ctx).unwrap();
        println!("Root seed item added.");
    }

    // 4. Launch Parallel Tasks
    let thread_count = 10; // Reduced from 20 for quicker demo
    let items_per_thread = 100;
    let mut handles = vec![];

    // We need to share the database handle as well
    let db = Arc::new(db);

    for i in 0..thread_count {
        let ctx_clone = Arc::clone(&ctx);
        let db_clone = Arc::clone(&db);
        let thread_id = i;

        let handle = thread::spawn(move || {
            let mut retry_count = 0;
            let mut committed = false;
            
            while !committed && retry_count < 10 {
                if retry_count > 0 {
                    println!("Thread {} retrying ({}/10)...", thread_id, retry_count);
                    thread::sleep(Duration::from_millis(100));
                }

                // Start Transaction
                let trans = match db_clone.begin_transaction(&ctx_clone) {
                    Ok(t) => t,
                    Err(e) => {
                        eprintln!("Thread {} failed to begin transaction: {}", thread_id, e);
                        retry_count += 1;
                        continue;
                    }
                };

                // Open Btree
                // Note: In a real app, you might cache the btree handle, but here we open it per transaction
                // to simulate independent clients.
                let btree = match db_clone.open_btree::<i64, String>(&ctx_clone, "concurrent_tree", &trans, None) {
                    Ok(b) => b,
                    Err(e) => {
                        eprintln!("Thread {} failed to open btree: {}", thread_id, e);
                        // If we can't open the btree, maybe we should rollback?
                        // But trans is local.
                        let _ = trans.rollback(&ctx_clone);
                        retry_count += 1;
                        continue;
                    }
                };

                // Add Items
                let mut batch = Vec::with_capacity(items_per_thread);
                for j in 0..items_per_thread {
                    let key = (thread_id as i64 * items_per_thread as i64) + j as i64;
                    let value = format!("Thread {} - Item {}", thread_id, j);
                    batch.push(Item::new(key, value));
                }

                // We use upsert instead of update to handle potential re-runs if needed, 
                // and to allow adding new items.
                match btree.upsert_batch(&ctx_clone, batch) {
                    Ok(_) => {},
                    Err(e) => {
                        eprintln!("Thread {} failed to add batch: {}", thread_id, e);
                        let _ = trans.rollback(&ctx_clone);
                        retry_count += 1;
                        continue;
                    }
                }

                // Commit
                match trans.commit(&ctx_clone) {
                    Ok(_) => {
                        committed = true;
                        println!("Thread {} committed successfully.", thread_id);
                    },
                    Err(e) => {
                        eprintln!("Thread {} failed to commit: {}", thread_id, e);
                        // Transaction is likely already failed on server side, but we can try rollback locally
                        let _ = trans.rollback(&ctx_clone);
                        retry_count += 1;
                    }
                }
            }
            
            if !committed {
                eprintln!("Thread {} failed after 10 retries.", thread_id);
            }
        });
        handles.push(handle);

        // Add sleep jitter to stagger thread starts (20-200ms)
        let nanos = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().subsec_nanos();
        let millis = 20 + (nanos % 381);
        thread::sleep(Duration::from_millis(millis as u64));
    }

    for handle in handles {
        handle.join().unwrap();
    }

    println!("All threads finished.");
    
    // Verify count
    let trans = db.begin_transaction(&ctx).unwrap();
    // We need to open the btree again to get a handle bound to this transaction?
    // Actually, the btree handle holds the ID. But the operations take the context.
    // Wait, the btree operations don't take the transaction explicitly in the Rust API?
    // Let's check lib.rs.
    // btree.add(&ctx, item) -> manageBtree(ctx.id, ...)
    // The transaction is associated with the Btree when it was created/opened?
    // In Go: manageDatabase(NewBtree) -> returns Btree ID.
    // The Btree object in Go is associated with the Transaction it was created with.
    // So yes, we need a new Btree handle for the new transaction.
    
    let btree = db.open_btree::<i64, String>(&ctx, "concurrent_tree", &trans, None).unwrap();
    
    match btree.count() {
        Ok(count) => println!("Total items in B-Tree: {}", count),
        Err(e) => eprintln!("Error getting count: {}", e),
    }
    
    let mut success_count = 0;
    for i in 0..thread_count {
        let key = i as i64 * items_per_thread as i64;
        match btree.find(&ctx, key) {
            Ok(found) => {
                if found {
                    success_count += 1;
                } else {
                    eprintln!("Verification failed: Key {} not found.", key);
                }
            },
            Err(e) => eprintln!("Verification error for key {}: {}", key, e),
        }
    }
    
    trans.commit(&ctx).unwrap();

    db.remove_btree(&ctx, "concurrent_tree").unwrap();
    
    println!("Verification: Found representative keys for {}/{} threads.", success_count, thread_count);
}

File: bindings/python/sop/examples/concurrent_demo.py
import sys
import os
import threading
import time
import random
import logging

# Add parent directory to path to import sop
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from sop.context import Context
from sop.database import Database, DatabaseOptions, DatabaseType
from sop.btree import Item
from sop.redis import Redis
from sop import Logger, LogLevel

def worker(ctx, db, store_name, thread_id, items_per_thread):
    # NOTE: No threading.Lock used here.
    # SOP handles transaction isolation and merging.
    
    retry_count = 0
    committed = False
    
    while not committed and retry_count < 10:
        try:
            # Each thread starts its own transaction
            t = db.begin_transaction(ctx)
            btree = db.open_btree(ctx, store_name, t)            
            
            batch = []
            for j in range(items_per_thread):
                # Unique keys per thread -> No conflicts, SOP merges changes
                key = (thread_id * items_per_thread) + j                
                batch.append(Item(key=key, value=f"Thread {thread_id} - Item {j}"))
            
            if not btree.add(ctx, batch):
                raise Exception(f"Thread {thread_id} failed to write batch (add returned False)")

            t.commit(ctx)
            committed = True
            print(f"Thread {thread_id} committed. Btree ID: {btree.id}")
        except Exception as e:
            retry_count += 1
            delay = random.randint(100, 500) / 1000.0 * retry_count
            print(f"Thread {thread_id} conflict detected (Retry {retry_count}): {e}")
            time.sleep(delay)
            
    if not committed:
        print(f"Thread {thread_id} failed after retries.")

def main():
    logging.basicConfig(level=logging.DEBUG)
    print("--- Concurrent Transactions Demo ---")
    print("Demonstrating multi-threaded access without client-side locks.")
    print("SOP handles ACID transactions, conflict detection, and merging.")
    print("This applies to both local threads and distributed processes (Clustered mode).")
    
    db_path = "data/concurrent_demo_py"
    
    store_name = "concurrent_tree"
    Redis.initialize("redis://localhost:6379")
    
    Logger.configure(LogLevel.Warn)
    
    ctx = Context()
    db = Database(DatabaseOptions(
        stores_folders=[db_path],
        type=DatabaseType.Clustered
    ))
    
    # 1. Setup
    # IMPORTANT: Pre-seed the B-Tree with one item to establish the root node.
    # This prevents race conditions on the very first commit when multiple threads 
    # try to initialize an empty tree simultaneously.
    # NOTE: This requirement is simply to have at least one item in the tree.
    # It can be a real application item or a dummy seed item.
    t_setup = db.begin_transaction(ctx)
    btree = db.new_btree(ctx, store_name, t_setup)
    btree.add(ctx, Item(key=-1, value="Root Seed Item"))
    t_setup.commit(ctx)
    
    threads = []
    thread_count = 30
    items_per_thread = 300
    
    print(f"Launching {thread_count} threads, {items_per_thread} items each...")
    
    for i in range(thread_count):
        t = threading.Thread(target=worker, args=(ctx, db, store_name, i, items_per_thread))
        threads.append(t)
        t.start()
        # Jitter to simulate more realistic real life scenario on transactions running in the cluster.
        time.sleep(random.randint(20, 500) / 1000.0)
        
    for t in threads:
        t.join()
        
    print("done")
        
    # Verify
    t_read = db.begin_transaction(ctx)
    btree = db.open_btree(ctx, store_name, t_read)
    print(f"Verify Btree ID: {btree.id}")
    
    count = 0
    if btree.first(ctx):
        while True:
            count += 1
            if not btree.next(ctx):
                break
            
    # Expected count includes the seed item
    expected_count = (thread_count * items_per_thread) + 1
    print(f"Final Count: {count} (Expected: {expected_count})")
    
    if count == expected_count:
        print("SUCCESS: All transactions merged correctly.")
    else:
        print("FAILURE: Count mismatch.")
        
    t_read.commit(ctx)

    db.remove_btree(ctx, store_name)
    Redis.close()

if __name__ == "__main__":
    main()
